### PR TITLE
`text/applyEdit` method

### LINKED ISSUE #549: Implement `text/applyEdit`
### Summary
The key operation of the text editing workflow is the ability for the client to inform the language server about a change that it has made. This underlies the whole functionality as it ensures that the language server (and other clients) keep their buffer representations in sync.

### Value
The clients and language server will all be able to stay in sync when changes are made to buffers.

### Specification

- [ ] Implement any auxiliary data structures necessary for passing this information.
- [ ] Implement the `text/applyEdit` message, that applies a list of file edits, potentially across multiple files, as specified in the [design document](https://github.com/luna/enso/blob/master/doc/design/engine/engine-services.md).
- [ ] If an edit is applied in an out of date fashion (the `oldVersion` UUID doesn't match), the edit must be rejected and return an error. 
- [ ] Document any errors in the design document that may be returned by this message.

### Acceptance Criteria & Test Cases

- The above specification has been implemented.
- The above functionality has been rigorously tested.

### PR DESCRIPTION
**Pull request by lolczak**
### Pull Request Description
The key operation of the text editing workflow is the ability of the client to inform the language server about a change that it has made. This underlies the whole functionality as it ensures that the language server (and other clients) keep their buffer representations in sync.

closes #549

### Important Notes
None

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: engine/language-server/src/main/scala/org/enso/languageserver/text/CollaborativeBuffer.scala [hunk 1]
```diff
@@ -104,6 +114,76 @@ class CollaborativeBuffer(
...
+  }
+
+  private def doEdit(
+    buffer: Buffer,
+    edits: List[TextEdit]
+  ): Either[ApplyEditFailure, Buffer] = {
+    EditorOps
+      .applyEdits(buffer.contents, edits)
+      .leftMap(toEditFailure)
+      .map(rope => Buffer(rope, versionCalculator.evalVersion(rope.toString)))
```
Comment (kustosz) [inline engine/language-server/src/main/scala/org/enso/languageserver/text/CollaborativeBuffer.scala]: Is it possible to use `CharSequence` as an argument for the version calculator? If so, you could use `rope.characters` and avoid reallocating the string here.

File: engine/language-server/src/main/scala/org/enso/languageserver/text/editing/RopeTextEditor.scala [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  override def edit(buffer: Rope, diff: TextEdit): Rope = {
+    val head: Rope = cutOutHead(buffer, diff)
+    val tail: Rope = cutOutTail(buffer, diff)
+
+    head ++ Rope(diff.text) ++ tail
+  }
+
+  private def cutOutHead(buffer: Rope, diff: TextEdit): Rope = {
+    val fullLines =
+      if (diff.range.start.line > 0)
```
Comment (kustosz) [inline engine/language-server/src/main/scala/org/enso/languageserver/text/editing/RopeTextEditor.scala]: Why is this necessary? Is there some issue with `rope.lines.take(0)`?

File: engine/language-server/src/test/scala/org/enso/languageserver/text/editing/RopeTextEditorSpec.scala [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                             |    result = plusOne 10
+                             |    result
+                             |    return result""".stripMargin
+  }
+
+  it should "support code points above 0xFFFF" in {
+    //given
+    //0x0001F4AF
+    val utf32Text          = Rope("unicode: \ud83d\udcaf end")
+    val positionInCodUnits = Range(Position(0, 9), Position(0, 10))
```
Comment (kustosz) [inline engine/language-server/src/test/scala/org/enso/languageserver/text/editing/RopeTextEditorSpec.scala]: should be `positionInCodeUnits`, I think?

Comment (lolczak) [reply engine/language-server/src/test/scala/org/enso/languageserver/text/editing/RopeTextEditorSpec.scala]: fixed

File: engine/language-server/src/main/scala/org/enso/languageserver/text/editing/RopeTextEditor.scala (hunk 2 above)
Comment (lolczak) [reply engine/language-server/src/main/scala/org/enso/languageserver/text/editing/RopeTextEditor.scala]: fixed

File: engine/language-server/src/main/scala/org/enso/languageserver/text/CollaborativeBuffer.scala (hunk 1 above)
Comment (lolczak) [reply engine/language-server/src/main/scala/org/enso/languageserver/text/CollaborativeBuffer.scala]: It has to be String in order to convert it to UTF-8 byte array easily.
